go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+		ObjectName:    &absKey,
+	}
+
+	res, err := bs.client.HeadObject(ctx, req)
+	if err != nil {
+		if serr, ok := common.IsServiceError(err); ok {
+			// handle not found code
+			if serr.GetHTTPStatusCode() == 404 {
+				return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+			}
+		}
+		return nil, "", err
+	}
+
+	requestedSize, totalSize, offSet, err := getDownloadInfo(br, res.ContentLength)
+	if err != nil {
+		return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+	}
+
+	etag := fmtstr(res.ETag)
+
+	return bs.multipartDownload(ctx, absKey, namespace, etag, requestedSize, totalSize, offSet)
+}
+
+func (bs *OCIBlobstore) multipartDownload(ctx context.Context, absKey, namespace, etag string, requestedSize, totalSize, offSet int64) (io.ReadCloser, string, error) {
+	parts, err := downloadParts(ctx, absKey, namespace, etag, requestedSize, totalSize, defaultBatchSize, offSet, defaultPartSize, maxPartNum, defaultConcurrentListeners, bs.downloadPart)
+	if err != nil {
+		return nil, "", err
+	}
+	rc, err := assembleParts(parts)
+	return rc, etag, err
+}
+
+func (bs *OCIBlobstore) getObject(ctx context.Context, absKey, namespace, etag string, br BlobRange) (io.ReadCloser, string, error) {
+	req := objectstorage.GetObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &absKey,
+	}
+
+	byteRange := br.asHttpRangeHeader()
+	if byteRange != "" {
+		req.Range = &byteRange
+	}
+
+	if etag != "" {
+		req.IfMatch = &etag
+	} else {
+		star := "*"
+		req.IfNoneMatch = &star
+	}
+
+	res, err := bs.client.GetObject(ctx, req)
+	if err != nil {
+		if serr, ok := common.IsServiceError(err); ok {
+			// handle not found code
+			if serr.GetHTTPStatusCode() == 404 {
+				return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+			}
+		}
+		return nil, "", err
+	}
+
+	// handle negative offset and positive length
+	if br.offset < 0 && br.length > 0 {
+		trimmedR, err := trimContent(res, br.length)
+		if err != nil {
+			return nil, "", err
+		}
+		return trimmedR, fmtstr(res.ETag), nil
+	}
+
+	return res.Content, fmtstr(res.ETag), nil
+}
+
+// Put sets the blob and the version for a key
+func (bs *OCIBlobstore) Put(ctx context.Context, key string, reader io.Reader) (string, error) {
+	return bs.upload(ctx, "", key, reader)
+}
+
+// CheckAndPut will check the current version of a blob against an expectedVersion, and if the
+// versions match it will update the data and version associated with the key
+func (bs *OCIBlobstore) CheckAndPut(ctx context.Context, expectedVersion, key string, reader io.Reader) (string, error) {
+	return bs.upload(ctx, expectedVersion, key, reader)
+}
+
+// At the time of this implementation, Oracle Cloud does not provide a way to create composite objects
+// via their APIs/SDKs.
+func (bs *OCIBlobstore) Concatenate(ctx context.Context, key string, sources []string) (string, error) {
+	panic("concatenate is unimplemented on the oci blobstore")
```

Comment:
Better to just return an error? I agree something is misconfigured if this is being called...

Or can we restructure the implementations and interfaces so that this isn't necessary here at all, and only the GCS blobstore is implementing a different interface that actually supports this and the table persister that works with that thing take that?